haproxy_stats.sh

```diff
@@ -169,41 +174,50 @@ query_stats() {
     fi
 }
 
-# generate stats cache file
-get_stats() {
-  find ${CACHE_STATS_FILEPATH} -mmin +${CACHE_STATS_EXPIRATION} -delete >/dev/null 2>&1
-  if [ ! -e ${CACHE_STATS_FILEPATH} ]
+# a generic cache management function, that relies on 'flock'
+check_cache() {
+  local cache_type="${1}"
+  local cache_filepath="${2}"
+  local cache_expiration="${3}"  
+  local cache_filemtime
+  if "${FLOCK_BIN}" --shared --wait "${FLOCK_WAIT}" 200
   then
-    debug "no cache file found, querying haproxy"
-    _TMPFILE=${CACHE_STATS_FILEPATH}.${RANDOM}.tmp
-    query_stats "show stat" > ${_TMPFILE}
-    if [ -f $FILE ];
-        then
-            mv -f ${_TMPFILE} ${CACHE_STATS_FILEPATH}
-        fi
-  else
-    debug "cache file found, results are at most ${CACHE_STATS_EXPIRATION} minutes stale.."
-  fi
+    cache_filemtime=$(stat -c '%Y' "${cache_filepath}" 2> /dev/null)
+    if [ $((cache_filemtime+60*cache_expiration)) -ge ${CUR_TIMESTAMP} ]
+    then
+      debug "${cache_type} file found, results are at most ${cache_expiration} minutes stale.."
+      return 0
+    fi
+  fi 200> "${cache_filepath}${FLOCK_SUFFIX}"
+  if "${FLOCK_BIN}" --exclusive --wait "${FLOCK_WAIT}" 200
+  then
+    cache_filemtime=$(stat -c '%Y' "${cache_filepath}" 2> /dev/null)
+    if [ $((cache_filemtime+60*cache_expiration)) -ge ${CUR_TIMESTAMP} ]
+    then
+      debug "${cache_type} file found, results have been just updated by another process.."
+    else
+      debug "no ${cache_type} file found, querying haproxy"
+      query_stats "show ${cache_type}" > "${cache_filepath}"
+    fi
+  fi 200> "${cache_filepath}${FLOCK_SUFFIX}"
+}
+
+# generate stats cache file if needed
+get_stats() {
+  check_cache 'stat' "${CACHE_STATS_FILEPATH}" ${CACHE_STATS_EXPIRATION}
```

Comment:
Well... The extra quotes are not necessary at all. It only makes difference if there are spaces in the path, which is rarely used.

However, I believe extra quoting would prevent [the famous bumblebee bug #123](https://github.com/MrMEEE/bumblebee-Old-and-abbandoned/issues/123) from happening. This is why I usually quote variables in my scripts. It is a kind of "better safe than sorry" behavior.